design/one-pager-managed-resource-api-design.md

```diff
@@ -0,0 +1,294 @@
+or user may not want some of their resources in their Crossplane environment.
+
+Ideally, for the resources that do have Crossplane representations, you can append `Ref` to the field name,
+i.e. `network (string)` -> `networkRef (Network)`
+
+In some cases, we might have collisions or cases where provider field name is too similar to another field in the
+managed resource's Custom Resource fields. The section [Owner-based Struct Tree](#owner-based-struct-tree) tries to
+tackle that.
+
+### High Fidelity
+
+Related to https://github.com/crossplaneio/crossplane/issues/621 and https://github.com/crossplaneio/crossplane/issues/530
+
+In most of the cases, an application setup requires the cloud resources to be connected/related in a custom way for
+various configurations. An example of this could be resource grouping which could happen as a physical layer
+(VPC network) or logical (resource grouping in providers). So, one can assume that in most of the cases where you have
+more than one service in your setup, you want to relate one to another, hence you need to _refer_ to one another in
+their configurations. There could be one or many configuration fields in any resource that we may not know beforehand.
+
+The easiest answer to solve this is probably just exposing everything the provider exposes. This is a good starting
+point but not a complete solution. A given resource can be either _referrer_ or _referred_ in a setup. In the former,
+we're talking about how we can refer to another resource in the `Spec` but for the latter, it's about how best this
+resource can be referred in another resource's `Spec`. There is also abundance of fields that are not necessary for either
+scenario but are necessary on a level that is higher than Crossplane, like creation timestamps on a Web UI built on top
+of Crossplane.
+
+Ideally, what we want is to expose all possible configurations in `Spec` and whatever might be needed by the output 
+users of this resource in the `Status`. Generally, it's better to just copy all fields from the provider's API as
+starting point and eliminate them one by one if needed.
+
+What goes into `Spec`:
+Does the provider allow configuration of this parameter at _any_ point in the life cycle of the resource? Include if yes.
+This includes the fields that are late-initialized, meaning some fields could be populated by the provider when you
+actually create the resource but it also gives you the ability to change them. An example for this could be
+auto-generated resource tags. If the provider tags the resource without us telling them to do so, controller should
+update `Spec` and input user should make changes on that current value of the field. Related to https://github.com/crossplaneio/crossplane-runtime/issues/10
+
+What goes into `Status`:
+Does the provider give the information of that field on the first level of the query, `GET`?
```

Comment:
Are you still thinking this would be a superset of the spec? i.e. That we'd include the current observed state of all spec fields in case they diverged from the spec? I think that's implied here, but I think it may be worth making it explicit given that it's an uncommon pattern.

(FWIW I'm still super on the fence about whether we should do this.)

Reply:
Ah - I see this is covered below in immutable properties but I still think it's worth pointing out out here.

Reply:
As long as we can come up with something else that covers the following cases, I'd be more than happy to change that.

* It should be sufficient to allow access to `Status` sub-resource(via RBAC) for all-purpose information extraction by other entities. Meaning, I would like to be able to extract the current settings, identification information, metadata etc. only by looking at `Status`.
* We need to represent the most up-to-date state in happy cases as well as error cases. If user changed the desired state and if we either in the middle of applying it/couldn't apply it/never plan to apply it, the user should be able to know get the current state even if no error shows up. If we have an update-able config only in `Spec`, we're not able to do that.

However, we can also discuss the need for two points mentioned here. For instance, if we decide the first one is not worth covering, we can remove the creation-time and metadata fields from `Status` and keep only update-able fields there.

Reply:
Give there's a solid technical reason to do this, but that it's an uncommon pattern for Kubernetes, I suggest we avoid doing it for the time being but make sure there's an issue tracking the idea. We can then implement this if the main problem we foresee (hidden drift between the desired and actual state) becomes a real issue based on user feedback.

Reply:
Opened https://github.com/crossplaneio/crossplane/issues/925